### PR TITLE
Update `EISMap` 

### PR DESCRIPTION
**Pull request by wtbarnes**
The first iteration of `EISMap` was written against an older version of sunpy. This refactors the `EISMap` source to take advantage of features in sunpy >= 3.1. It also avoids making modifications to the metadata in place and instead overrides properties of `GenericMap` to accomplish the same task.

Additionally, this pins to sunpy >= 4.0 as releases older than this are now unsupported (e.g. there will be no bugfix releases for versions older than 4.0).

A few other minor fixes:

- The `measurement` property now just returns the thing being measured (e.g. intensity, velocity)
- The `nickname` property has been overridden to 'Hinode EIS <line_id> <measurement>'
- Make pinning for `sphinx` less restrict. This is needed as sphinx < 4.2 is not compatible with Python 3.10 (see sphinx-doc/sphinx#9562

**This PR should be squashed before merging**

### CODE REVIEW

**Comment by PaulJWright** [comment]
This is a significant change which I believe removes support for python 3.7, and sunpy < 4.0.

If this is going to be merged, I would prefer this was done, tested, and updated before the completion of the JOSS review. This may however be out of scope. I will consult with @mbobra and @nabobalis in the review thread: https://github.com/openjournals/joss-reviews/issues/4914

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -13,9 +13,9 @@ jobs:
       fail-fast: false
       matrix:
         include:
-        - python-version: 3.7
         - python-version: 3.8
         - python-version: 3.9
+        - python-version: '3.10'
```
Comment (nabobalis) [inline .github/workflows/tests.yml:18]: Do you want to also try 3.11?

Comment (nabobalis) [reply .github/workflows/tests.yml:18]: This should also be updated in the setup.cfg.

**Comment by MJWeberg** [comment]
Thanks @wtbarnes, for updating this! Since we got burned by dependency updates in the past, I did some extra testing in an isolated environment (not doubting your code, just trying to be extra careful in case something else in the dependency chain broke).

Everything looks great! I shall go ahead and merge it.  

@nabobalis, I believe I saw something recently about NDCube still testing Python 3.11 support. Since this is one of our main dependencies, would you recommend we start testing 3.11 ourselves, or wait until the next NDCube update?

**Comment by PaulJWright** [comment]
@MJWeberg before this is merged you may want to check that it updates the docs etc reflecting the versions

**Comment by nabobalis** [comment]
> @nabobalis, I believe I saw something recently about NDCube still testing Python 3.11 support. Since this is one of our main dependencies, would you recommend we start testing 3.11 ourselves, or wait until the next NDCube update?

It was just me testing that ndcube works under 3.11. From what I recall, there was no problems for ndcube.

Your issue for 3.11 might be that there are no wheels for h5py which means you need to compile them on the CI.

**Pull request merged**